### PR TITLE
Copy `cortex/pkg/ruler` package dependency into Loki

### PR DESCRIPTION
**Pull request by DylanGuedes**
**What this PR does / why we need it**:
As a part of removing Loki dependency on cortex packages, we decided to copy the `ruler` package from cortex to Loki itself.
Note that I decided to copy it into a `base` subpackage instead of merging with existing `ruler` package to make reviewing this easier. If you prefer that I try merging them into a single package at this same PR, just let me know.

**Which issue(s) this PR fixes**:
N/A

**Special notes for your reviewer**:
- Note that I decided to copy it into a `base` subpackage instead of merging it with the existing `ruler` package to make reviewing this easier. If you prefer that I try merging them into a single package at this same PR, just let me know.

**Checklist**
- [ ] Documentation added
- [ ] Tests updated
- [ ] Add an entry in the `CHANGELOG.md` about the changes.

### CODE REVIEW

**Comment by DylanGuedes** [comment]
I recommend reviewing this commit-by-commit.

**Review by jeschkies — APPROVED** [review]

File: pkg/ruler/rulestore/objectclient/rule_store.go [hunk 1]
```diff
@@ -8,15 +8,15 @@ import (
 	"io/ioutil"
 	"strings"
 
+	"github.com/cortexproject/cortex/pkg/chunk"
```
Comment (cyriltovena) [inline pkg/ruler/rulestore/objectclient/rule_store.go:11]: Replace with the Loki one.

Comment (DylanGuedes) [reply pkg/ruler/rulestore/objectclient/rule_store.go]: good catch. pushed a few commits changing it, wdyt?

File: pkg/loki/config_wrapper.go [hunk 2]
```diff
@@ -8,7 +8,6 @@ import (
 	"time"
 
 	cortexcache "github.com/cortexproject/cortex/pkg/chunk/cache"
```
Comment (cyriltovena) [inline pkg/loki/config_wrapper.go:10]: can this be loki ?

File: pkg/ruler/base/manager_metrics.go [hunk 3]
```diff
@@ -1,4 +1,4 @@
-package ruler
+package base
 
 import (
 	"github.com/prometheus/client_golang/prometheus"
```
Comment (cyriltovena) [inline pkg/ruler/base/manager_metrics.go:4]: The util package below could be avoided ?

File: pkg/ruler/base/notifier.go [hunk 4]
```diff
@@ -1,4 +1,4 @@
-package ruler
+package base
 
 import (
 	"context"
```
Comment (cyriltovena) [inline pkg/ruler/base/notifier.go:4]: Same for the util package of cortex below ?

File: pkg/ruler/base/notifier_test.go [hunk 5]
```diff
@@ -0,0 +1,235 @@
+package base
+
+import (
+	"fmt"
+	"testing"
+	"time"
+
+	"github.com/cortexproject/cortex/pkg/util"
```
Comment (cyriltovena) [inline pkg/ruler/base/notifier_test.go:8]: same we should get those function in loki.

File: pkg/loki/config_wrapper.go (hunk 2 above)
Comment (DylanGuedes) [reply pkg/loki/config_wrapper.go:10]: Just checked - we can't until we merge the `queryrange` fork.

File: pkg/ruler/base/manager_metrics.go (hunk 3 above)
Comment (DylanGuedes) [reply pkg/ruler/base/manager_metrics.go:4]: Nope, unfortunately. However, we are already working on forking `util`.

**Review by cyriltovena — APPROVED** [review]
LGTM

**Pull request merged**